### ISSUE TITLE
[Bug] `model.outcome_transform` does not update after running `model.condition_on_observations()` in `SingleTaskGP`

### ISSUE THREAD

**Issue by ptighin** [pytorch/botorch#1239]
## Issue description

I was surprised to find out that the outcome transform of `SingleTaskGP` is only trained once in the constructor and never re-trained when the model is updated with new data. This is a big problem when doing Bayesian optimization, where the model is updated at each step. Since the normalizer is not updated, the training data of the model is not normalized either. Is this a bug or the intended behavior?

## Code example

The following test nicely reproduces the issue:

```python
import pytest
import torch
from botorch.models import SingleTaskGP
from botorch.models.transforms import Standardize

def test_normalization_singletaskgp():
    X = torch.randn((3, 1))
    Y = torch.sin(X)

    model = SingleTaskGP(X, Y, outcome_transform=Standardize(1))

    # This passes
    assert model.train_targets.mean(dim=-1).item() == pytest.approx(0.0, abs=1e-6)
    assert model.train_targets.std(dim=-1).item() == pytest.approx(1.0, abs=1e-6)

    # `condition_on_observations` can't be run without doing model inference first
    with torch.no_grad():
        model.posterior(torch.zeros(0, 1))

    X_additional = torch.randn((12, 1))
    Y_additional = torch.sin(X_additional)
    model = model.condition_on_observations(X_additional, Y_additional)

    # This fails
    assert model.train_targets.mean(dim=-1).item() == pytest.approx(0.0, abs=1e-6)
    assert model.train_targets.std(dim=-1).item() == pytest.approx(1.0, abs=1e-6)
```

**Comment by Balandat** [comment]
Nice catch, but this is not a bug, at least not *really*. 

Note that `condition_on_observations` is not actually refitting the model parameters (e.g. length scales etc.), it just updates the data and leaves the inferred model parameters fixed. In general we use this functionality for things like "fantasizing", where we condition on "fake" observations during computations of look-ahead acquisition functions. It should not really be used in a standard BO loop where model fitting times are usually dominated by the time required for optimizing the acquisition function.

So if you instead just instantiate a new model with the appended data the transforms will properly learn the new standardization factors (if you do want to warm-start the optimization from previous model parameter values to speed up the fitting, you can use the `state_dict()` and `load_state_dict()` functionality to do so).

While one could change the behavior to have `condition_on_observations` also update the transforms, this seems unnecessary (and would also slow things down) given how we typically use this functionality. Note also that while the normalization factors will not be exact, if there is a reasonable amount of data already and the new training points / observations are drawn from a similar distribution as the existing ones, then they won't change too much and should be reasonable - importantly, the un-normalization and un-standardization will still work correctly, and the model should also work well even if the data isn't precisely but only approximately normalized / standardized.

**Comment by ptighin** [comment]
@Balandat thanks a lot for the insightful reply. I understand your reasoning. I have several follow-up questions.
> Note also that while the normalization factors will not be exact, if there is a reasonable amount of data already and the new training points / observations are drawn from a similar distribution as the existing ones, then they won't change too much and should be reasonable - importantly, the un-normalization and un-standardization will still work correctly, and the model should also work well even if the data isn't precisely but only approximately normalized / standardized.

This makes sense, although things may not be as well-behaved at the beginning of the BO loop when few data points are present. Do you think that this approach is still justified in this regime?

It seems to me that standardizing the data _without_ refitting the model parameters may bring one or two benefits:
* The model data does not suffer from finite bias.
* `condition_on_observations()` could also be used at every BO step to update the model (refitting the parameters can be done separately during likelihood optimization), This can make other models inheriting from `SingleTaskGP`'s parents quite more flexible, since instantiating the model at each BO loop places relatively strict constraints on the design choice of these models.

If I understand correctly, the downside of this approach is the slowdown due to the prediction caching, since the Woodbury vector and the Cholesky decomposition of the covariance matrix would need to be re-cached from scratch. I do not have a feeling about the practical impact of this slowdown and how much it balances the other two benefits.

**Comment by Balandat** [comment]
> This makes sense, although things may not be as well-behaved at the beginning of the BO loop when few data points are present. Do you think that this approach is still justified in this regime?

You’re correct that in the low data regime this approach will be a lot more problematic. In an optimization loop I would always recommend to fit a new model (incl. the learnable transform parameters) if possible (potentially warm-starting from the previous stare dict). This usually isn’t feasible for the „fantasizing“ we do within each evaluation of lookahead acquisition functions.


> It seems to me that standardizing the data without refitting the model parameters

I haven’t really thought about this, but it seems that this approach could potentially be quite problematic in that this would change all of the training data of the model and so the MLL would be optimized w.r.t potentially quite different data. Whereas for the approach where we add data without renormalizing we’d have all the same data points except for the additionally added one. As I said above haven’t really thought through this, but at least my gut feeling is that renormalizing without refitting could cause more things to go wrong. 

Curious if @j-wilson or @bletham have any more thoughts on this.

**Comment by ptighin** [comment]
> So if you instead just instantiate a new model with the appended data the transforms will properly learn the new standardization factors (if you do want to warm-start the optimization from previous model parameter values to speed up the fitting, you can use the `state_dict()` and `load_state_dict()` functionality to do so).

I've thought a little about this and the problem seems to be that if I instantiate a new model with the appended data and _afterwards_ load the previous model parameters, the transforms will be "undone" and set to the old values, isn't it? So one would have, after the model instantiation and `load_state_dict()`, to train the transform yet again. Do I understand this correctly or am I missing something? It seems to be a little cumbersome but definitely doable.

> I haven’t really thought about this, but it seems that this approach could potentially be quite problematic in that this would change all of the training data of the model and so the MLL would be optimized w.r.t potentially quite different data. Whereas for the approach where we add data without renormalizing we’d have all the same data points except for the additionally added one. As I said above haven’t really thought through this, but at least my gut feeling is that renormalizing without refitting could cause more things to go wrong.

If my understanding is correct, the key advantage of not doing data renormalisation is that one can cache the new Cholesky decomposition much faster by "warm-starting" from the old Cholesky. In case the data changes this no longer works and the whole cache needs to be recalculated from scratch. I don't know how much speedup this offers in practice but if it's significant, I definitely understand the logic of doing things this way.

**Comment by Balandat** [comment]
> I've thought a little about this and the problem seems to be that if I instantiate a new model with the appended data and afterwards load the previous model parameters, the transforms will be "undone" and set to the old values, isn't it? So one would have, after the model instantiation and load_state_dict(), to train the transform yet again. Do I understand this correctly or am I missing something? It seems to be a little cumbersome but definitely doable.

You are correct on what will happen upon loading the state dict. But the goal with this approach is not to avoid re-fitting the model altogether. By "warm-start the optimization" I mean that since we're starting from the previously optimal parameters (incl. the standardization parameters) the optimization should (and typically does) converge much faster. Presumably the standardization parameters in the `state_dict` will only be slightly off from the actual ones, so this should usually be ok. If you really don't like this approach you could also do first remove the standardization parameters from the state dict, and then use `load_state_dict` with `strict=False`.  

> If my understanding is correct, the key advantage of not doing data renormalisation is that one can cache the new Cholesky decomposition much faster by "warm-starting" from the old Cholesky. In case the data changes this no longer works and the whole cache needs to be recalculated from scratch. I don't know how much speedup this offers in practice but if it's significant, I definitely understand the logic of doing things this way.

Indeed, the speedups can be very significant, and this is the whole idea behind having `condition_on_observations`.

**Comment by ptighin** [comment]
> If you really don't like this approach you could also do first remove the standardization parameters from the state dict, and then use load_state_dict with strict=False.

Thanks for the tip, @Balandat, it is exactly what one would need, particularly at the beginning of the optimisation when a new data point can significantly change the standardisation parameters. I will close this issue since my questions were answered -- thank you!

**Comment by Fr3574** [comment]
Hi @Balandat, I stumbled upon this issue as I encountered an issue in my BO-loop since `Standardize` now became the [default](https://github.com/pytorch/botorch/blob/main/CHANGELOG.md#0120----sep-17-2024) for `outcome_transform`. Shouldn't either the `outcome_transform.xxx`-parameters be removed entirely from the `state_dict` or refitted by `fit_gpytorch_mll`?

If you start with a bad estimate for `outcome_transform._stdvs_sq`, e.g. 1 if you only have a single training point, then this will never change if you overwrite the dict with the one from the previous run.

My current workaround:
```
model = GP( ...)
if state_dict is not None:
  state_dict_wo_outcome_transform = {k: v for k, v in state_dict.items() if not k.startswith("outcome_transform")}
  model.load_state_dict(state_dict_wo_outcome_transform, strict=False)
mll = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_mll(mll)
```

Happy to hear to your opinion on this :)

**Comment by Balandat** [comment]
@Fr3574 sorry just saw this

> Shouldn't either the outcome_transform.xxx-parameters be removed entirely from the state_dict or refitted by fit_gpytorch_mll?

Are these parameters not being refitted by `fit_gpytorch_mll`? I think they should be, no?
